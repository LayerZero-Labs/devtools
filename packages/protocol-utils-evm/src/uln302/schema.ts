import { AddressSchema } from '@layerzerolabs/utils'
import { BigNumberishBigintSchema } from '@layerzerolabs/utils-evm'
import { z } from 'zod'

/**
 * Schema for parsing an ethers-specific UlnConfig into a common format
 */
export const Uln302UlnConfigSchema = z.object({
    confirmations: BigNumberishBigintSchema,
    requiredDVNs: z.array(AddressSchema),
    optionalDVNs: z.array(AddressSchema),
    optionalDVNThreshold: z.coerce.number().int().nonnegative(),
})

/**
 * Schema for parsing a common UlnConfig into a ethers-specific format
 */
export const Uln302UlnConfigInputSchema = Uln302UlnConfigSchema.transform((config) => ({
    ...config,
    confirmations: String(config.confirmations),
    requiredDVNCount: config.requiredDVNs.length,
    optionalDVNCount: config.optionalDVNs.length,
}))
