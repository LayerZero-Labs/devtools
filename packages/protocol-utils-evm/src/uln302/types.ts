import type { BigNumberish } from '@ethersproject/bignumber'
import { Uln302ExecutorConfig, Uln302UlnConfig } from '@layerzerolabs/protocol-utils'

export interface Uln302UlnConfigInput extends Omit<Uln302UlnConfig, 'confirmations'> {
    confirmations: BigNumberish
}

export interface Uln302ExecutorConfigInput extends Omit<Uln302ExecutorConfig, 'maxMessageSize'> {
    maxMessageSize: BigNumberish
}
