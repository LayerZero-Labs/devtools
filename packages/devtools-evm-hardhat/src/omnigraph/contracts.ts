import pMemoize from 'p-memoize'
import { ProviderFactory, connectOmniContract } from '@layerzerolabs/devtools-evm'
import { createContractFactory } from '@/omnigraph/coordinates'
import type { OmniContractFactoryHardhat } from '@/omnigraph/types'
import { createProviderFactory } from '@/provider'

export const createConnectedContractFactory = (
    contractFactory: OmniContractFactoryHardhat = createContractFactory(),
    providerFactory: ProviderFactory = createProviderFactory()
): OmniContractFactoryHardhat =>
    pMemoize(async (point) => {
        const contract = await contractFactory(point)
        const provider = await providerFactory(point.eid)

        return connectOmniContract(contract, provider)
    })
