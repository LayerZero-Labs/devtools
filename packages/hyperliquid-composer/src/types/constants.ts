export const HYPERLIQUID_URLS = {
    MAINNET: 'https://api.hyperliquid.xyz',
    TESTNET: 'https://api.hyperliquid-testnet.xyz',
}

export const RPC_URLS = {
    MAINNET: 'https://rpc.hyperliquid.xyz/evm',
    TESTNET: 'https://rpc.hyperliquid-testnet.xyz/evm',
}

export const CHAIN_IDS = {
    MAINNET: 999,
    TESTNET: 998,
}

export const ENDPOINTS = {
    INFO: '/info',
    EXCHANGE: '/exchange',
}

export const MAX_HYPERCORE_SUPPLY = 2 ** 64 - 1

/**
 * These are the token ids for USDC on hypercore mainnet and testnet
 * It can be reproduced by grabbing the entire spotMeta and finding USDC because Hyperliquid does not have an API to target query an asset (April 8, 2025)
 *  -   curl -X POST "https://api.hyperliquid.xyz/info" -H "Content-Type: application/json" -d '{"type": "spotMeta"}' > spotOut.json
 *  -   Goto tokens and the first entry should be USDC with an index of 0
 */
export const USDC_TOKEN_ID = {
    MAINNET: 0,
    TESTNET: 0,
}

export function toAssetBridgeAddress(tokenIndex: number): string {
    const addressLength = 42
    const addressPrefix = '0x2'
    const indexAsHex = Number(tokenIndex).toString(16)

    const addressLengthWithoutPrefix = addressLength - addressPrefix.length
    return `${addressPrefix}${indexAsHex.padStart(addressLengthWithoutPrefix, '0')}`
}
