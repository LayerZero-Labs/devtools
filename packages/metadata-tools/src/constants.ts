export const METADATA_URL = process.env.LZ_METADATA_URL || 'https://metadata.layerzero-api.com/v1/metadata'
