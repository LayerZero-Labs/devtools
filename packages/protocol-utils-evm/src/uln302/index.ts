export * from './factory'
export * from './sdk'
export * from './types'
