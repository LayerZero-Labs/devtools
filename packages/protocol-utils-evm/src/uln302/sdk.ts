import type { EndpointId } from '@layerzerolabs/lz-definitions'
import type { IUln302, Uln302ExecutorConfig, Uln302UlnConfig } from '@layerzerolabs/protocol-utils'
import { Address, formatEid, type OmniTransaction } from '@layerzerolabs/utils'
import { omniContractToPoint, type OmniContract, makeZeroAddress } from '@layerzerolabs/utils-evm'
import { Uln302ExecutorConfigSchema, Uln302UlnConfigInputSchema, Uln302UlnConfigSchema } from './schema'

export class Uln302 implements IUln302 {
    constructor(public readonly contract: OmniContract) {}

    async getUlnConfig(eid: EndpointId, address?: Address | null | undefined): Promise<Uln302UlnConfig> {
        const config = await this.contract.contract.getUlnConfig(makeZeroAddress(address), eid)

        // Now we convert the ethers-specific object into the common structure
        //
        // Here we need to spread the config into an object because what ethers gives us
        // is actually an array with extra properties
        return Uln302UlnConfigSchema.parse({ ...config })
    }

    async getExecutorConfig(eid: EndpointId, address?: Address | null | undefined): Promise<Uln302ExecutorConfig> {
        const config = await this.contract.contract.getExecutorConfig(makeZeroAddress(address), eid)

        // Now we convert the ethers-specific object into the common structure
        //
        // Here we need to spread the config into an object because what ethers gives us
        // is actually an array with extra properties
        return Uln302ExecutorConfigSchema.parse({ ...config })
    }

    async setDefaultExecutorConfig(eid: EndpointId, config: Uln302ExecutorConfig): Promise<OmniTransaction> {
        const data = this.contract.contract.interface.encodeFunctionData('setDefaultExecutorConfigs', [
            [{ eid, config }],
        ])

        return this.createTransaction(data)
    }

    async setDefaultUlnConfig(eid: EndpointId, config: Uln302UlnConfig): Promise<OmniTransaction> {
        const serializedConfig = Uln302UlnConfigInputSchema.parse(config)
        const data = this.contract.contract.interface.encodeFunctionData('setDefaultUlnConfigs', [
            [
                {
                    eid,
                    config: serializedConfig,
                },
            ],
        ])

        return {
            ...this.createTransaction(data),
            description: `Setting default ULN config for ${formatEid(eid)}: ${JSON.stringify(serializedConfig)}`,
        }
    }

    protected createTransaction(data: string): OmniTransaction {
        return {
            point: omniContractToPoint(this.contract),
            data,
        }
    }
}
