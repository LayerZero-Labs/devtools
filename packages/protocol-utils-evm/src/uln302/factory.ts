import pMemoize from 'p-memoize'
import type { OmniContractFactory } from '@layerzerolabs/utils-evm'
import type { Uln302Factory } from '@layerzerolabs/protocol-utils'
import { Uln302 } from './sdk'

/**
 * Syntactic sugar that creates an instance of EVM `Uln302` SDK
 * based on an `OmniPoint` with help of an `OmniContractFactory`
 *
 * @param {OmniContractFactory} contractFactory
 * @returns {Uln302Factory<Uln302>}
 */
export const createUln302Factory = <TOmniPoint = never>(
    contractFactory: OmniContractFactory<TOmniPoint>
): Uln302Factory<Uln302, TOmniPoint> => pMemoize(async (point) => new Uln302(await contractFactory(point)))
