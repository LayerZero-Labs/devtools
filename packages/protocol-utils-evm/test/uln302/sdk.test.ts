import { type OmniContract } from '@layerzerolabs/utils-evm'
import { Uln302 } from '@/uln302'
import { MainnetEndpointId } from '@layerzerolabs/lz-definitions'
import { Contract } from '@ethersproject/contracts'
import { AddressZero } from '@ethersproject/constants'
import { Uln302ExecutorConfig, Uln302UlnConfig } from '@layerzerolabs/protocol-utils'
import artifact from '@layerzerolabs/lz-evm-sdk-v2/artifacts/contracts/uln/uln302/SendUln302.sol/SendUln302.json'

describe('uln302/sdk', () => {
    let contract, omniContract: OmniContract, ulnSdk

    beforeAll(async () => {
        contract = new Contract(AddressZero, artifact.abi)
        omniContract = { eid: MainnetEndpointId.ETHEREUM_MAINNET, contract }
        ulnSdk = new Uln302(omniContract)
    })

    describe('encodeExecutorConfig', () => {
        it('should encode and decode the Uln302ExecutorConfig', async () => {
            const executorConfig: Uln302ExecutorConfig = { executor: AddressZero, maxMessageSize: 100 }
            const executorConfigEncoded = ulnSdk.encodeExecutorConfig(executorConfig)
            expect(executorConfigEncoded).toMatchSnapshot()

            const executorConfigDecoded: Uln302ExecutorConfig = ulnSdk.decodeExecutorConfig(executorConfigEncoded)
            expect(executorConfig).toEqual(executorConfigDecoded)
        })
    })

    describe('encodeUlnConfig', () => {
        it('should encode and decode the Uln302UlnConfig', async () => {
            const ulnConfig: Uln302UlnConfig = {
                confirmations: BigInt(100),
                optionalDVNThreshold: 1,
                optionalDVNs: [AddressZero, AddressZero],
                requiredDVNs: [AddressZero],
            }
            const ulnConfigEncoded = ulnSdk.encodeUlnConfig(ulnConfig)
            expect(ulnConfigEncoded).toMatchSnapshot()

            const ulnConfigDecoded: Uln302ExecutorConfig = ulnSdk.decodeUlnConfig(ulnConfigEncoded)

            expect(ulnConfig).toEqual(ulnConfigDecoded)
        })
    })
})
