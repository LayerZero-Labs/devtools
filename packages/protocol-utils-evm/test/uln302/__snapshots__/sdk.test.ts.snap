// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uln302/sdk encodeExecutorConfig should encode and decode the Uln302ExecutorConfig 1`] = `"0x00000000000000000000000000000000000000000000000000000000000000640000000000000000000000000000000000000000000000000000000000000000"`;

exports[`uln302/sdk encodeUlnConfig should encode and decode the Uln302UlnConfig 1`] = `"0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000006400000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;
